pkg/kstatus/watcher/watcher_dynamic.go

```diff
@@ -0,0 +1,825 @@
+					Namespace(namespace).
+					List(ctx, options)
+			},
+			WatchFunc: func(options metav1.ListOptions) (watch.Interface, error) {
+				return w.DynamicClient.Resource(mapping.Resource).
+					Namespace(namespace).
+					Watch(ctx, options)
+			},
+		},
+		example,
+		w.ResyncPeriod,
+		cache.Indexers{
+			cache.NamespaceIndex: cache.MetaNamespaceIndexFunc,
+		},
+	)
+
+	// Handler called when ListAndWatch errors.
+	// Custom handler stops the informer if the resource is NotFound (CRD deleted).
+	err = informer.SetWatchErrorHandler(func(r *cache.Reflector, err error) {
+		w.watchErrorHandler(groupKind, err)
+	})
+	if err != nil {
+		// Should never happen.
+		// Informer can't have started yet. We just created it.
+		panic(fmt.Sprintf("Failed to set error handler on new informer for %v: %v", mapping.Resource, err))
+	}
+
+	eventCh := make(chan event.Event)
+
+	// Add this event channel to the output multiplexer
+	w.muxer.Input(eventCh)
+
+	informer.AddEventHandler(w.eventHandler(ctx, eventCh))
+
+	// Start the informer in the background.
+	// Informer will be stopped when the context is cancelled.
+	go func() {
+		klog.V(3).Infof("Watch of %v starting", groupKind)
+		informer.Run(ctx.Done())
+		klog.V(3).Infof("Watch of %v stopped", groupKind)
+		// Signal to the caller there will be no more events for this GroupKind.
+		close(eventCh)
+	}()
+
+	return nil
+}
+
+// readStatusFromObject is a convenience function to read object status with a
+// StatusReader using a ClusterReader to retrieve generated objects.
+func (w *ObjectStatusInformer) readStatusFromObject(
+	ctx context.Context,
+	obj *unstructured.Unstructured,
+) (*event.ResourceStatus, error) {
+	return w.StatusReader.ReadStatusForObject(ctx, w.ClusterReader, obj)
+}
+
+// readStatusFromCluster is a convenience function to read object status with a
+// StatusReader using a ClusterReader to retrieve the object and its generated
+// objects.
+func (w *ObjectStatusInformer) readStatusFromCluster(
+	ctx context.Context,
+	id object.ObjMetadata,
+) (*event.ResourceStatus, error) {
+	return w.StatusReader.ReadStatus(ctx, w.ClusterReader, id)
+}
+
+// deletedStatus builds a ResourceStatus for a deleted object.
+//
+// StatusReader.ReadStatusForObject doesn't handle nil objects as input. So
+// this builds the status manually.
+// TODO: find a way to delegate this back to the status package.
+func deletedStatus(id object.ObjMetadata) *event.ResourceStatus {
+	// Status is always NotFound after deltion.
+	// Passed obj represents the last known state, not the current state.
+	result := &event.ResourceStatus{
+		Identifier: id,
+		Status:     status.NotFoundStatus,
+		Message:    "Resource not found",
+	}
+
+	return &event.ResourceStatus{
+		Identifier: id,
+		Resource:   nil, // deleted object has no
+		Status:     result.Status,
+		Message:    result.Message,
+		// Presumably if the object is deleted, so are all the resources it generated.
```

Comment:
This is only true if resources are deleted using foreground cascading deletion: https://kubernetes.io/docs/concepts/architecture/garbage-collection/#foreground-deletion

Reply:
True, but the Poller also has the same behavior in the `baseStatusReader`. If lookupResource returns NotFound, it builds a ResourceStatus with nil GeneratedResources.

I changed the comment to be more clear, but I don't think the behavior is a problem for our use case, because if deletion policy is NOT configured to be foreground, the desired behavior for confirming deletion implies ignoring the status of generated resources. So I left a TODO to improve it for more general use, if that becomes a priority.